pkg/tempo/instance.go

```diff
@@ -141,11 +142,19 @@ func (i *Instance) buildAndStartPipeline(ctx context.Context, cfg InstanceConfig
 	}
 
 	if cfg.SpanMetrics != nil && len(cfg.SpanMetrics.PromInstance) != 0 {
-		prom, err := promManager.GetInstance(cfg.SpanMetrics.PromInstance)
-		if err != nil {
-			return fmt.Errorf("failed to get prometheus instance %s", cfg.SpanMetrics.PromInstance)
+		for {
+			select {
+			case <-ctx.Done():
+				return fmt.Errorf("failed to get prometheus instance %s", cfg.SpanMetrics.PromInstance)
+			default:
+				prom, err := promManager.GetInstance(cfg.SpanMetrics.PromInstance)
+				if err == nil {
+					ctx = context.WithValue(ctx, contextkeys.Prometheus, prom)
+					break
+				}
+				<-time.After(defaultRetryInterval)
```

Comment:
Instead of looping when building the pipeline, what do you think about passing the whole manager to the span metrics processor and perform the look up of the instance per span? It should be a fairly speedy lookup.

Reply:
Good idea, made an implementation in 944efc7b160c2867d88e85ea806d3f15d9718e81.

I think losing a few metrics during start up is reasonable

Reply:
That commit looks good, but did you mean to push it to this PR? I'm not seeing it here, only via that link.

Reply:
I pushed it to the main repository, my bad. It should be fixed now!